src/util/heap/layout/vm_layout_constants.rs

```diff
@@ -120,48 +34,175 @@ pub const LOG_PAGES_IN_SPACE64: usize = LOG_SPACE_SIZE_64 - LOG_BYTES_IN_PAGE as
- * conditional definition.
- */
-#[cfg(target_pointer_width = "32")]
-pub const SPACE_MASK_64: usize = 0;
-#[cfg(target_pointer_width = "64")]
-pub const SPACE_MASK_64: usize = ((1 << LOG_MAX_SPACES) - 1) << SPACE_SHIFT_64;
-
-/*
- * Size of each space in the 64-bit memory layout
- *
- * We can't express this constant in a 32-bit environment, hence the
- * conditional definition.
- */
-// FIXME: When Compiling for 32 bits this expression makes no sense
-// #[allow(const_err)]
-// pub const SPACE_SIZE_64: usize = if_then_else_usize!(HEAP_LAYOUT_64BIT,
-//    1 << LOG_SPACE_SIZE_64, MAX_SPACE_EXTENT);
-#[cfg(target_pointer_width = "64")]
-pub const SPACE_SIZE_64: usize = 1 << LOG_SPACE_SIZE_64;
-#[cfg(target_pointer_width = "32")]
-pub const SPACE_SIZE_64: usize = MAX_SPACE_EXTENT;
+impl VMLayoutConstants {
+    /// Normal 32-bit configuration
+    pub const fn new_32bit() -> Self {
+        Self {
+            log_address_space: 32,
+            heap_start: chunk_align_down(unsafe { Address::from_usize(0x8000_0000) }),
+            heap_end: chunk_align_up(unsafe { Address::from_usize(0xd000_0000) }),
+            vm_space_size: chunk_align_up(unsafe { Address::from_usize(0xdc0_0000) }).as_usize(),
+            log_max_chunks: Self::LOG_ARCH_ADDRESS_SPACE - LOG_BYTES_IN_CHUNK,
+            log_space_extent: 31,
+            space_shift_64: 0,
+            space_mask_64: 0,
+            space_size_64: 1 << 31,
+        }
+    }
+    /// Normal 64-bit configuration
+    #[cfg(target_pointer_width = "32")]
+    pub fn new_64bit() -> Self {
+        unimplemented!("64-bit heap constants do not work with 32-bit builds")
+    }
+    #[cfg(target_pointer_width = "64")]
+    pub fn new_64bit() -> Self {
+        Self {
+            log_address_space: 47,
+            heap_start: chunk_align_down(unsafe {
+                Address::from_usize(0x0000_0200_0000_0000usize)
+            }),
+            heap_end: chunk_align_up(unsafe { Address::from_usize(0x0000_2200_0000_0000usize) }),
+            vm_space_size: chunk_align_up(unsafe { Address::from_usize(0xdc0_0000) }).as_usize(),
+            log_max_chunks: Self::LOG_ARCH_ADDRESS_SPACE - LOG_BYTES_IN_CHUNK,
+            log_space_extent: 41,
+            space_shift_64: 41,
+            space_mask_64: ((1 << 4) - 1) << 41,
+            space_size_64: 1 << 41,
+        }
+    }
+    /// 64-bit configuration with compressed pointers
+    pub fn new_64bit_with_pointer_compression(heap_size: usize) -> Self {
+        assert!(
+            heap_size <= (32usize << LOG_BYTES_IN_GBYTE),
+            "Heap size is larger than 32 GB"
+        );
+        let start = 0x4000_0000;
+        let end = match start + heap_size {
+            end if end <= (4usize << 30) => 4usize << 30,
+            end if end <= (32usize << 30) => 32usize << 30,
+            _ => 0x4000_0000 + (32usize << 30),
+        };
+        Self {
+            log_address_space: 35,
```

Comment:
Pointer compression doesn't necessarily have to be 35-bit address space. Android uses 32-bit address space. I was using an older version of this patch and set the address space to 32-bits here: https://github.com/k-sareen/mmtk-core/blob/b68292cbe8c875906cf19501c34641628c9fbc3c/src/util/heap/layout/vm_layout_constants.rs#L111

Reply:
The new `VMLayoutConstants::set_custom_vm_layout_constants` interface should solve this as well. Bindings can specify a different `log_address_space`

Reply:
I don't think we want to expose the entire `VMLayoutConstants` to the binding. The struct includes many fields that we do not want the binding to customize, and we cannot guarantee MMTk to work with arbitrary values from the bindings. I suggest exposing limited options through our runtime `Option`, such as `use_compressed_pointer_bits`. We only need to expose options that we are supporting right now. If we haven't tested things like `heap_start`, we do not need to expose them at this point.

Reply:
runtime `Option` is user configurable, this is not a good place for VM configurable values. For most bindings there is no need to use this function, as the default 32bit or 64bit values should be enough. For bindings that use this function, they should make sure the heap constants are correct.

Also, I believe make VMs be able to configure `heap_start` is necessary. You can't just assume our default fixed `heap_start` can work for all the VMs without any virtual space conflict. I actually did solve a virtual address conflict issue on GoogleJDK by manually modifying `heap_start` in mmtk-core. If VM bindings has no control to heap start address, this will happen again to other VMs.

Reply:
> runtime `Option` is user configurable, this is not a good place for VM configurable values. For most bindings there is no need to use this function, as the default 32bit or 64bit values should be enough. For bindings that use this function, they should make sure the heap constants are correct.

If runtime option is not suitable, we can use constants on `VMBinding`. My point is we should only expose values that we want the bindings to customize.

> Also, I believe make VMs be able to configure `heap_start` is necessary. You can't just assume our default fixed `heap_start` can work for all the VMs without any virtual space conflict. I actually did solve a virtual address conflict issue on GoogleJDK by manually modifying `heap_start` in mmtk-core. If VM bindings have no control to heap start address, this will happen again to other VMs.

Yeah. We can expose `heap_start` if you tested that our code base works with an arbitrary heap start.

Reply:
> if you tested that our code base works with an arbitrary heap start.

This can never be true. e.g. zero address can never work. I think it's VM binding's responsibility to ensure their configured heap range works with their virtual memory.

Reply:
I agree with Wenyu here. It's up to the binding implementer to check that the heap start actually works for them. I had to mess around with the heap start for Android as well. But I think it makes sense to only expose a limited set of options/functions to the binding so that there is less chance of using stuff incorrectly.